oak_kernel_measurement/src/main.rs

```diff
@@ -0,0 +1,123 @@
+// limitations under the License.
+//
+
+use std::path::PathBuf;
+
+use anyhow::Context;
+use clap::Parser;
+use sha2::{Digest, Sha256};
+use zerocopy::FromBytes;
+
+/// The default workspace-relative path to the Linux Kernel bzImage file.
+const DEFAULT_LINUX_KERNEL: &str = "oak_containers_kernel/target/bzImage";
+
+#[derive(Parser, Clone)]
+#[command(about = "Oak Kernel Measurement Calculator")]
+struct Cli {
+    #[arg(long, help = "The location of the kernel bzImage file")]
+    kernel: Option<PathBuf>,
+}
+
+impl Cli {
+    fn kernel_path(&self) -> PathBuf {
+        self.kernel.clone().unwrap_or_else(|| {
+            format!("{}/{}", env!("WORKSPACE_ROOT"), DEFAULT_LINUX_KERNEL).into()
+        })
+    }
+}
+
+fn main() -> anyhow::Result<()> {
+    env_logger::init();
+    let cli = Cli::parse();
+    let kernel_info = Kernel::load(cli.kernel_path()).context("couldn't load kernel file")?;
+    let mut kernel_hasher = Sha256::new();
+    kernel_hasher.update(&kernel_info.kernel_image);
+    println!(
+        "Kernel Image Measurement: sha2-256:{}",
+        hex::encode(kernel_hasher.finalize())
+    );
+
+    let mut setup_hasher = Sha256::new();
+    setup_hasher.update(&kernel_info.setup_data);
+    println!(
+        "Kernel Setup Data Measurement: sha2-256:{}",
+        hex::encode(setup_hasher.finalize())
+    );
+
+    Ok(())
+}
+
+struct Kernel {
+    setup_data: Vec<u8>,
+    kernel_image: Vec<u8>,
+}
+
+impl Kernel {
+    /// Parses a bzImage kernel file and extracts the kernel image and setup data.
+    ///
+    /// The VMM will parse the the bzImage file and split it into two parts: the setup data (which
+    /// includes the real-mode code) and the 64-bit kernel image. The VMM also updates some fields
+    /// in the setup data header.
+    ///
+    /// See <https://github.com/qemu/qemu/blob/6630bc04bccadcf868165ad6bca5a964bb69b067/hw/i386/x86.c#L795>
+    /// and <https://www.kernel.org/doc/html/v6.7/arch/x86/boot.html>
+    fn from_bz_image(bz_image: &[u8]) -> Self {
+        // The number of 512 byte sectors -1 is stored at offset 0x1F1.
+        let setup_sects = bz_image[0x1F1];
+        // For backwards compatibility, if setup_sects is 0 it should be set to 4.
+        let setup_sects = if setup_sects == 0 { 4 } else { setup_sects };
+        let setup_size = 512 * (setup_sects as usize + 1);
+
+        // The kernel image is just everything after the setup data.
+        let kernel_image = bz_image[setup_size..].to_vec();
+        let mut setup_data = bz_image[..setup_size].to_vec();
+        // The VMM will modify some fields in the setup data header.
+        // The loader type will be set to `QEMU`.
+        setup_data[0x210] = 0xB0;
+        // The load flags will be updated to enable heap usage.
+        setup_data[0x211] |= 0x80;
+        // Set the default command_line location.
+        let default_cmd_line = 0x20000;
+        *u32::mut_from(&mut setup_data[0x228..0x22C]).expect("invalid slice for cmd_line") =
```

Comment:
Are these addresses always the same? Like, the same for all Linux kernels?

Reply:
Yes, they are specified in the Linux boot protocol (https://www.kernel.org/doc/html/v6.7/arch/x86/boot.html). Newer versions might add new fields, but the offsets for existing fields don't change.